### PR TITLE
Data sync protected

### PR DESCRIPTION
**Pull request by aliok**
https://issues.jboss.org/browse/AEROGEAR-8294

### CODE REVIEW

File: packages/sync/src/links/HeadersLink.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  private headerProvider: HeaderProvider;
+
+  constructor(headerProvider: HeaderProvider) {
+    super();
+    this.headerProvider = headerProvider;
+  }
+
+  public request(operation: Operation, forward: NextLink) {
+    if (this.headerProvider) {
```
Comment (wtrocki) [inline packages/sync/src/links/HeadersLink.ts]: Redundant.

File: packages/sync/src/config/HeaderProvider.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Interface for HeaderProvider
+ */
+export interface HeaderProvider {
+  getHeaders(): {[index: string]: string};
```
Comment (wtrocki) [inline packages/sync/src/config/HeaderProvider.ts]: [Optional]`{[index: string]: string};`  can be defined as type

**Comment by wtrocki** [comment]
We are missing the main part of auth SDK logic that has `refreshToken` done and it's called inside users app.

**Comment by aliok** [comment]
> Looks good to me. The only question is should we have another HeaderProvider implementation that will add the authorisation header for users that are logged in via the auth module?

> We are missing the main part of auth SDK logic that has refreshToken logic and it's called inside users app.

Why don't we leave that to the user? Otherwise I don't see why we're generalizing things as a `HeaderProvider`.
In Passos's work, we had a `requiresAuth` flag which was to make everything automatic if set. In this new case though, an implementation has to be provided and it is not targeted to be made automatic.

However, I don't know the background of why we're changing in approach. @wtrocki  can give more info perhaps.

See the sample app modification here: https://github.com/aerogear/apollo-voyager-ionic-example/commit/55a9a248e45d94a2293bac5bd20f02eac9c9239b#diff-92c003e539499632c72be26749e89db4R27
PS: sample app is not working good at the moment and the code I wrote there is far from being ideal.

**Pull request merged**